Restrict reparenting of mission-affiliated carried ships to the same missison

tehhowch opened the pull request:
Refs #4307 

Also reduces the frequency with which carried ships that are escorts of a ship that cannot carry them seek different parents.

petervdmeer commented on source/AI.cpp:
```diff
@@ -702,29 +702,59 @@ void AI::Step(const PlayerInfo &player)
-							parent = other;
-							it->SetParent(other);
+							auto send = n.Ships().end();
+							return send != find_if(n.Ships().begin(), send, [&it](const shared_ptr<Ship> &s)
```
This part looks very magical to me. Could use some comments on what is happening here.

tehhowch replied:
This set of lambdas is finding the associated mission, if any, in which this ship belongs to one of the NPC blocks, using `std::find_if`.

tehhowch commented on source/AI.cpp:
```diff
@@ -702,29 +702,59 @@ void AI::Step(const PlayerInfo &player)
 				parentChoices.reserve(ships.size() * .1);
-				for(const auto &other : ships)
-					if(other->GetGovernment() == gov && other->GetSystem() == it->GetSystem() && !other->CanBeCarried())
+				auto reparentWith = [&](const list<shared_ptr<Ship>> otherShips) -> bool
```
The functionality in this lambda isn't needed anywhere else, even within the AI class, except within this block. For that reason, I chose to use a lambda over an anonymous namespace method or a class method

petervdmeer replied:
Okay,  but could you then at least then put a comment just above the lambda function to mention that a lamba function is used here? I know a lot of programmers that are unfamiliar with lambda functions and if such a programmer sees this for the first time in his/her life, then I would like such programmer to at least have a pointer/indication of what is happening here.

tehhowch commented:
Still need to verify that this fixes the root cause of #4307 before it merges.
I might also add caching of the mission iterator; I need to look at how frequently this code actually gets executed to determine if that has merit.

tehhowch commented:
Added a guard against a special ship that has no system (which is what causes the segfault in #4307)
Also removed a couple of the nested lambdas in favor of a Mission class method, which should help readability.
It's possible that caching the mission iterator should also be done; adding all accepted missions from the "ManyMissions Tester" save file in #4424 to the save file provided in 4307 jumps the in-flight CPU counter for my laptop from 2% to 20-22% -- I will make a decision on that tomorrow.

Feel free to provide an additional review @petervdmeer

petervdmeer commented on source/AI.cpp:
```diff
@@ -702,29 +702,52 @@ void AI::Step(const PlayerInfo &player)
+							break;
+				}
+				else
+					reparentWith(ships);
```
Is it possible that this line of code causes a non-special/non-mission fighter to be reparented to a special/mission carrier? And if that happens, would that be bad? (For example due to the carrier no longer having space for a mission fighter?)

petervdmeer replied:
As discussed on discord: not an issue, solved in CanCarry method
